### PR TITLE
Bee ring implementation needs

### PR DESCRIPTION
**Pull request by biotinker**
This PR is the first chunk of code necessary to get our external partner Bee Ring supported on stock RDK.

There are two pieces of this PR:
1) adds support for the `sht3xd` temperature sensor
2) adds support for having sensor data upload to the "Sensor Data" part of our Data Management system

Since I don't usually make PRs that touch so many different bits of RDK  feel free to add anyone else to the PR if I missed them/they should take a look

### CODE REVIEW

File: components/sensor/sht3xd/sht3xd.go
Comment (mcvella) [inline components/sensor/sht3xd/sht3xd.go:169]: While we do not really have full standards here as far as I know, we should strive to be consistent on units etc(partial list here: https://viam.atlassian.net/wiki/spaces/ENG/pages/262371/Viam+Engineering+Guidelines)

So, I think humidity_pct ?
On the temp ones, I don't think we have a convention and the d318b20 temp sensor just returns "temp" in C.  Maybe we should decide on a convention even if it means a breaking change now for d318b20?
@Fahmina

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [inline components/sensor/sht3xd/sht3xd.go:22]: [nit] is it useful to have the manufacturer in the model name for this chip?

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [inline components/sensor/sht3xd/sht3xd.go:48]: [nit] we like having the field match the attribute field exactly, so `i2c_bus`, in this case.

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [inline components/sensor/sht3xd/sht3xd.go:2]: [minor] add datasheet link, and any relevant repos we should be looking at if ever troubleshooting this component

could find [this](https://cdn-shop.adafruit.com/product-files/2857/Sensirion_Humidity_SHT3x_Datasheet_digital-767294.pdf), but want to confirm that was what you were looking at before adding to the rdk drive.

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [inline components/sensor/sht3xd/sht3xd.go:106]: [minor] the datasheet and examples I looked at included information about setting a polling time, and a time constant for I guess reading the sensor. Would it be best to run that as a background goroutine here rather than open up the bus and close it each time we read?

The poll time examples I've seen from a cursory google search are around 60s, but can be configured. @agreenb, any poll time concerns on data's side?

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [inline components/sensor/sht3xd/sht3xd.go:151]: [q] confirming this would show up as an empty map on app and an error card popping up? Will this client be primarily using app to interface with us or programming their own logic?

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [reply components/sensor/sht3xd/sht3xd.go:169]: agree with `humidity_pct` 

I prefer sticking with SI, so centigrade scale makes sense to me.

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [inline components/sensor/sht3xd/sht3xd.go:168]: [q] is there a reason to return both? Does Bee Ring want to operate in F?

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [inline components/sensor/sht3xd/sht3xd.go:175]: [q] datasheet I looked at prescribes a minimum of 350 ns for sensor resets on this chip. From my understanding, I think what you're doing here is sufficient to not hit that case, but want to be conscious of this fact if you see this reset logic failing a lot.

**Review by randhid — CHANGES_REQUESTED** [review]
Done with first pass! A few minor things, can you give a bit more context on the use case by so we can focus on a good experience with this sensor addition?

File: components/sensor/collector.go
Comment (agreenb) [inline components/sensor/collector.go:49]: Neat, I like this functionality. I want to be consistent in how we're passing in reader names between the sensor and board components. Do you think a user will typically want to just grab _all_ readers?

If so, I'll update the board collector similarly — if `len(arg) == 0`, grab all `AnalogReaderNames()` and create a reader per name. cc @npmenard who added the board collector.

File: components/sensor/collector.go
Comment (agreenb) [inline components/sensor/collector.go:43]: If there are any readers that require `extra` parameters, should support grabbing (and validating that is a map type?) `args['extra']` if non-nil. Can make this into a ticket to be nonblocking for this PR.

File: components/sensor/sht3xd/sht3xd.go
Comment (agreenb) [reply components/sensor/sht3xd/sht3xd.go:106]: We've set frequencies of up to 16kHz on arm and gantry, but this was a while back and will be benchmarking again soon. Each reading is kicked off in its own goroutine so not blocked by a separate sensor reading — along with each reading, we also store the time the reading was requested and when it was received. It's mostly constrained by the hardware.

File: components/sensor/collector.go
Comment (biotinker) [reply components/sensor/collector.go:43]: I was thinking about that, the reason I decided not to for this PR was that this would interact with a sensor that returned a reading whose name was "extra"; I didn't want to be the one making that call since this PR reaches outside my usual scope of code changes.

But if that's the way that you think this should work I'm happy to include in this PR.

File: components/sensor/collector.go
Comment (biotinker) [reply components/sensor/collector.go:49]: I do indeed think a user will generally want to grab all readers; I certainly did. IMO having the "Data Capture Configuration" field attached to a component means _a priori_ "I want data from this component", and if the user cares enough about specifics and granularity then they can add that too. If I add another analog or digital reader to my board I'll likely want that captured too.

Also in the case where there's one analog or one digital reader, and the board has a data capture config of any kind, that's pretty clear what's desired.

Comment (biotinker) [reply components/sensor/collector.go:49]: Though I would agree that the structure between sensor and board should be identical here; so if it's decided not to update board then this should change to mimic that.

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [reply components/sensor/sht3xd/sht3xd.go:169]: I'm not picky either, will ask product for you.

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [reply components/sensor/sht3xd/sht3xd.go:22]:
```suggestion
var modelname = resource.NewDefaultModel("sensiron-sht3xd")
```

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [inline components/sensor/sht3xd/sht3xd.go:1]:
```suggestion
// Package sht3xd implements a sht3x-d sensor for temperature and humidity
// datasheet can be found at: https://cdn-shop.adafruit.com/product-files/2857/Sensirion_Humidity_SHT3x_Datasheet_digital-767294.pdf
// repo:
```

File: components/sensor/sht3xd/sht3xd.go
Comment (biotinker) [reply components/sensor/sht3xd/sht3xd.go:106]: In high repeatability/accuracy mode, this sensor can require up to 15ms for the operation, so that would max out at 66hz.

I cannot think of any use case where someone would be using this specific sensor, and would want temperature data at 66hz; I'm not convinced that the thermal characteristics of the chip would be such that such a high temperature resolution would be meaningful.

If you'd prefer that this be updated to use a background goroutine (similar to what the ultrasonic sensor uses?) I can do that, but as written it's similar to the bme280 and ds18b20 sensors we have.

File: components/sensor/sht3xd/sht3xd.go
Comment (biotinker) [reply components/sensor/sht3xd/sht3xd.go:151]: Current plan is that the client will be building a bidirectional remote where all of their own logic will live. We're less interested in any particular reading, more interested in aggregating this data over long periods of time.

I do not believe that our `app` interface at present is sufficient for what they want, but we may work with app in the future to implement that if it's close.

File: components/sensor/sht3xd/sht3xd.go
Comment (biotinker) [reply components/sensor/sht3xd/sht3xd.go:168]: The `bme280` returns both so this matches that. But I can remove.

File: components/sensor/sht3xd/sht3xd.go
Comment (biotinker) [reply components/sensor/sht3xd/sht3xd.go:175]: I'll add a sleep.

File: components/sensor/sht3xd/sht3xd.go
Comment (mcvella) [inline components/sensor/sht3xd/sht3xd.go:25]:
```suggestion
var modelname = resource.NewDefaultModel("sensirion-sht3xd")
```
fixes misspelling

**Review by mcvella — CHANGES_REQUESTED** [review]

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [reply components/sensor/sht3xd/sht3xd.go:25]: thanks for catching that!

File: components/sensor/collector.go
Comment (agreenb) [reply components/sensor/collector.go:43]: I want to redesign how we're currently passing in reader names (e.g. support nested `additional_params`), and will include this in that refactor. If you could create a ticket for this and link to it in the TODO, that'd be helpful. Thank you!

File: components/sensor/collector.go
Comment (agreenb) [reply components/sensor/collector.go:49]: I would keep this functionality as-is, and I created https://viam.atlassian.net/browse/DATA-1104 to update accordingly in board.

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [reply components/sensor/sht3xd/sht3xd.go:109]: Let's go ahead and leave it as is for now. If we need to change it into a background goroutine (movement or ultrasonic sensors) and a polling time coming in from data or its own config, we'll do that as it becomes necessary.

**Review by agreenb — APPROVED** [review]
LGTM for the collector. Nice!

File: components/sensor/sht3xd/sht3xd.go
Comment (randhid) [inline components/sensor/sht3xd/sht3xd.go:171]: Let's do: 
```suggestion
		"relative_humidity_pct":     humid, \\ TODO: RSDK-1903
```
if you'd rather have something shorter would be open to rel_humidity_pct. Also created the corresponding ticket.

Comment (randhid) [reply components/sensor/sht3xd/sht3xd.go:171]: talked to @Fahmina , she likes my suggestion. Will update the wiki

File: components/sensor/collector.go
Comment (biotinker) [reply components/sensor/collector.go]: Done

**Review by randhid — APPROVED** [review]
LGTM

**Review by mcvella — APPROVED** [review]

**Pull request merged**
